sql: panic in makeDistSQLReceiver

This is on the release-2.0 branch.

```
panic: sync/atomic: store of inconsistently typed value into Value [recovered]
        panic: sync/atomic: store of inconsistently typed value into Value

goroutine 1046293 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4207f2120, 0x3550940, 0xc458cf1440)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:180 +0x136
panic(0x20301e0, 0x26a9a40)
        /home/jordan/go1.9/src/runtime/panic.go:491 +0x283
sync/atomic.(*Value).Store(0xc462d80068, 0x227a600, 0xc451718fc0)
        /home/jordan/go1.9/src/sync/atomic/value.go:79 +0x10d
github.com/cockroachdb/cockroach/pkg/sql.makeDistSQLReceiver.func1(0x352e340, 0xc451718fc0)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:340 +0x4d
github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).cleanupTxnLocked(0xc45c562540, 0x3550940, 0xc458cf15c0, 0x2)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:916 +0x1f5
github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).tryAsyncAbort(0xc45c562540, 0x3550940, 0xc458cf15c0)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:1019 +0x15f
github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).heartbeat(0xc45c562540, 0x3550940, 0xc458cf15c0, 0x0)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:1130 +0x667
github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).heartbeatLoop(0xc45c562540, 0x3550940, 0xc458cf15c0)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:990 +0x34f
github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).updateState.func1(0x3550940, 0xc458cf1440)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:1308 +0x3e
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc4207f2120, 0x3550940, 0xc458cf1440, 0xc497ee8d20, 0x29, 0x3561460, 0xc43cb0ba00, 0xc49c1aa8b0)
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:266 +0xe6
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
        /home/jordan/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:261 +0x133
```

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -250,6 +250,7 @@ type distSQLReceiver struct {
 	// canceled is atomically set to an error when this distSQL receiver
 	// has been marked as canceled. Upon the next Push(), err is set to
 	// this value, and ConsumerClosed is the ConsumerStatus.
+	// It contains errWraps only.
 	canceled atomic.Value
 
 	row    tree.Datums
@@ -271,6 +272,12 @@ type distSQLReceiver struct {
 	// A handler for clock signals arriving from remote nodes. This should update
 	// this node's clock.
 	updateClock func(observedTs hlc.Timestamp)
+}
+
+// errWrap is a container for an error, for use with atomic.Value, which
+// requires that all of things stored in it must have the same concrete type.
+type errWrap struct {
+	err error
 }
 
 // rowResultWriter is a subset of CommandResult to be used with the
@@ -337,7 +344,7 @@ func makeDistSQLReceiver(
 	// aborted, or committed), ensure the dist SQL flow is canceled.
 	if r.txn != nil {
 		r.txn.OnFinish(func(err error) {
-			r.canceled.Store(err)
+			r.canceled.Store(errWrap{err: err})
 		})
 	}
 	return r
@@ -401,7 +408,7 @@ func (r *distSQLReceiver) Push(
 	}
 	if r.resultWriter.Err() == nil && r.canceled.Load() != nil {
 		// Set the error to reflect query cancellation.
-		r.resultWriter.SetError(r.canceled.Load().(error))
+		r.resultWriter.SetError(r.canceled.Load().(errWrap).err)
 	}
 	if r.resultWriter.Err() != nil {
 		// TODO(andrei): We should drain here.
@@ -460,7 +467,7 @@ func (r *distSQLReceiver) ProducerDone() {
 
 // SetCanceled is part of the CancellableRowReceiver interface.
 func (r *distSQLReceiver) SetCanceled() {
-	r.canceled.Store(sqlbase.NewQueryCanceledError())
+	r.canceled.Store(errWrap{err: sqlbase.NewQueryCanceledError()})
 }
 
 // updateCaches takes information about some ranges that were mis-planned and

